dpcpp/solver/idr_kernels.dp.cpp

```diff
@@ -55,11 +66,760 @@ namespace dpcpp {
 namespace idr {
 
 
+constexpr int default_block_size = 256;
+constexpr int default_dot_dim = 16;
+constexpr int default_dot_size = default_dot_dim * default_dot_dim;
+
+
+template <typename ValueType>
+void initialize_m_kernel(size_type subspace_dim, size_type nrhs,
+                         ValueType *__restrict__ m_values, size_type m_stride,
+                         stopping_status *__restrict__ stop_status,
+                         sycl::nd_item<3> item_ct1)
+{
+    const auto global_id = thread::get_thread_id_flat(item_ct1);
+    const auto row = global_id / m_stride;
+    const auto col = global_id % m_stride;
+
+    if (global_id < nrhs) {
+        stop_status[global_id].reset();
+    }
+
+    if (row < subspace_dim && col < nrhs * subspace_dim) {
+        m_values[row * m_stride + col] =
+            (row == col / nrhs) ? one<ValueType>() : zero<ValueType>();
+    }
+}
+
+template <typename ValueType>
+void initialize_m_kernel(dim3 grid, dim3 block, size_t dynamic_shared_memory,
+                         sycl::queue *stream, size_type subspace_dim,
+                         size_type nrhs, ValueType *m_values,
+                         size_type m_stride, stopping_status *stop_status)
+{
+    stream->submit([&](sycl::handler &cgh) {
+        cgh.parallel_for(
+            sycl_nd_range(grid, block), [=](sycl::nd_item<3> item_ct1) {
+                initialize_m_kernel(subspace_dim, nrhs, m_values, m_stride,
+                                    stop_status, item_ct1);
+            });
+    });
+}
+
+
+template <size_type block_size, typename ValueType>
+void orthonormalize_subspace_vectors_kernel(
+    size_type num_rows, size_type num_cols, ValueType *__restrict__ values,
+    size_type stride, sycl::nd_item<3> item_ct1,
+    UninitializedArray<ValueType, block_size> *reduction_helper_array,
+    remove_complex<ValueType> *reduction_helper_real)
+{
+    const auto tidx = thread::get_thread_id_flat(item_ct1);
+
+
+    ValueType *__restrict__ reduction_helper = (*reduction_helper_array);
+
+
+    for (size_type row = 0; row < num_rows; row++) {
+        for (size_type i = 0; i < row; i++) {
+            auto dot = zero<ValueType>();
+            for (size_type j = tidx; j < num_cols; j += block_size) {
+                /*
+                DPCT1007:5: Migration of this CUDA API is not supported by the
+                Intel(R) DPC++ Compatibility Tool.
+                */
+                dot += values[row * stride + j] * conj(values[i * stride + j]);
+            }
+            // TODO: check with intel why we need this here.
+            item_ct1.barrier();
```

Comment:
I also think it should not need this one, but it is the place I can make it correct on cpu.
I do not have time to investigate it yet

Reply:
I am thinking it requires the global space sync because we update the value in global space later and use it in next loop.
Originally, I thought it always accessed by the same thread (with same tid), so it should be okay even it is not visible from global. However, it may not be true in CPU?

Reply:
We have two other barriers before we access `values `again, and the column is being used exclusively by that thread. Does the code break if you remove this synchronization? If so, that sounds like a compiler issue.

Reply:
Yes, it will break if I remove the barrier.
I hope it's a compile issue. In my understanding, the value should be update because it is on the same thread.
It is only happened in cpu and we also need this weird kind barrier in coo or csr, too.